File: bindings/lua/Polycode/Vector4.lua
class "Vector4"


function Vector4:__getvar(name)
	if name == "x" then
		return Polycode.Vector4_get_x(self.__ptr)
	elseif name == "y" then
		return Polycode.Vector4_get_y(self.__ptr)
	elseif name == "z" then
		return Polycode.Vector4_get_z(self.__ptr)
	elseif name == "w" then
		return Polycode.Vector4_get_w(self.__ptr)
	end
end

function Vector4:__setvar(name,value)
	if name == "x" then
		Polycode.Vector4_set_x(self.__ptr, value)
		return true
	elseif name == "y" then
		Polycode.Vector4_set_y(self.__ptr, value)
		return true
	elseif name == "z" then
		Polycode.Vector4_set_z(self.__ptr, value)
		return true
	elseif name == "w" then
		Polycode.Vector4_set_w(self.__ptr, value)
		return true
	end
	return false
end
function Vector4:Vector4(...)
	local arg = {...}
	for k,v in pairs(arg) do
		if type(v) == "table" then
			if v.__ptr ~= nil then
				arg[k] = v.__ptr
			end
		end
	end
	if self.__ptr == nil and arg[1] ~= "__skip_ptr__" then
		self.__ptr = Polycode.Vector4(unpack(arg))
	end
end

function Vector4:set(x, y, z, w)
	local retVal = Polycode.Vector4_set(self.__ptr, x, y, z, w)
end

function Vector4:dot(u)
	local retVal = Polycode.Vector4_dot(self.__ptr, u.__ptr)
	return retVal
end

function Vector4:__delete()
	if self then Polycode.delete_Vector4(self.__ptr) end
end

File: bindings/javascript/Polycode/Vector4.js
function Vector4(x,y,z,w) {
	if(arguments[0] != "__skip_ptr__") {
		this.__ptr = Polycode.Vector4(x,y,z,w)
	}
	Object.defineProperties(this, {
		'x': { enumerable: true, configurable: true, get: Vector4.prototype.__get_x, set: Vector4.prototype.__set_x},
		'y': { enumerable: true, configurable: true, get: Vector4.prototype.__get_y, set: Vector4.prototype.__set_y},
		'z': { enumerable: true, configurable: true, get: Vector4.prototype.__get_z, set: Vector4.prototype.__set_z},
		'w': { enumerable: true, configurable: true, get: Vector4.prototype.__get_w, set: Vector4.prototype.__set_w}
	})
}

Vector4.prototype.__get_x = function() {
	return Polycode.Vector4__get_x(this.__ptr)
}

Vector4.prototype.__set_x = function(val) {
	Polycode.Vector4__set_x(this.__ptr, val)
}

Vector4.prototype.__get_y = function() {
	return Polycode.Vector4__get_y(this.__ptr)
}

Vector4.prototype.__set_y = function(val) {
	Polycode.Vector4__set_y(this.__ptr, val)
}

Vector4.prototype.__get_z = function() {
	return Polycode.Vector4__get_z(this.__ptr)
}

Vector4.prototype.__set_z = function(val) {
	Polycode.Vector4__set_z(this.__ptr, val)
}

Vector4.prototype.__get_w = function() {
	return Polycode.Vector4__get_w(this.__ptr)
}

Vector4.prototype.__set_w = function(val) {
	Polycode.Vector4__set_w(this.__ptr, val)
}

Duktape.fin(Vector4.prototype, function (x) {
	if (x === Vector4.prototype) {
		return;
	}
	Polycode.Vector4__delete(x.__ptr)
})

Vector4.prototype.set = function(x,y,z,w) {
	Polycode.Vector4_set(this.__ptr, x, y, z, w)
}

Vector4.prototype.dot = function(u) {
	return Polycode.Vector4_dot(this.__ptr, u)
}
